### DISCUSSION TITLE
Why is NTP and lack thereof so important to Tasmota functionality?

### DISCUSSION THREAD

**Discussion by code-issues-pull** [arendst/Tasmota#14604] [Q&A] [answered]
### PROBLEM DESCRIPTION
**Here's my biggest issue with this... nowhere in the documentation is NTP mentioned as being important to basic functionality.** No timers or anything else are used when these problems happen. If Tasmota is denied NTP access it throws an absolute fit. And it doesn't make sense. It's not as if it's not able to track the passage of time internally even if the time itself doesn't match with the rest of the world (1970) it can still count when a second has passed.

My personal issues range from intermittent wifi dropouts to delayed response of commands sent over MQTT and HTML. All of these issues are immediately resolved by setting TimeZone and NtpServer1 to a server it can access. No other changes to devices or network and suddenly everything performs as expected with no further wifi drop outs.

Why is the lack of NTP access so detrimental to the functionality of some Tasmota devices? Why does wifi drop out frequently? Why are commands to them delayed? Why do they spam NTP requests (I'm basing this off my poor memory of DNS log activity from them when I first got them, 8.x firmware)?

More and more people are wanting to isolate their IOT networks, the default behavior for devices shouldn't be wifi dropouts and extremely delayed command responses if no NTP server is available. Shouldn't the basic functionality such as issuing a command be as simple as "oh, command to turn on received, perform action"?

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [X] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): USW-WF04 wifi outlets
- [X] Tasmota binary firmware version number used: Tasmota 10.1.0
  - [X] Pre-compiled
  - [ ] Self-compiled
- [X] Flashing tools used: web ota for the last 2 major versions
- [X] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here: 17:43:20.870 MQT: stat/tasmota_8487E4/RESULT = {"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO4":{"0":"None"},"GPIO16":{"0":"None"}}


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [X] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
17:46:42.675 MQT: stat/tasmota_8487E4/STATUS = {"Status":{"Module":0,"DeviceName":"Tas1","FriendlyName":["Tas1"],"Topic":"tasmota_8487E4","ButtonTopic":"0","Power":1,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
17:46:42.681 MQT: stat/tasmota_8487E4/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"5T02:15:30","StartupUTC":"2022-01-20T23:31:12","Sleep":50,"CfgHolder":4617,"BootCount":40,"BCResetTime":"2021-06-30T02:46:10","SaveCount":3920,"SaveAddress":"F9000"}}
17:46:42.686 MQT: stat/tasmota_8487E4/STATUS2 = {"StatusFWR":{"Version":"10.1.0(tasmota)","BuildDateTime":"2021-12-08T14:47:33","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"347/699"}}
17:46:42.690 MQT: stat/tasmota_8487E4/STATUS3 = {"StatusLOG":{"SerialLog":0,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["IOT","IOT"],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A000000000000","00000080","00006000","00004000"]}}
17:46:42.701 MQT: stat/tasmota_8487E4/STATUS4 = {"StatusMEM":{"ProgramSize":616,"Free":384,"Heap":22,"ProgramFlashSize":1024,"FlashSize":1024,"FlashChipId":"1440A1","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45","Sensors":"1,2,3,4,5,6"}}
17:46:42.709 MQT: stat/tasmota_8487E4/STATUS5 = {"StatusNET":{"Hostname":"0001","IPAddress":"10.0.1.11","Gateway":"10.0.1.1","Subnetmask":"255.255.255.0","DNSServer1":"10.0.0.5","DNSServer2":"0.0.0.0","Mac":"EC:FA:BC:84:87:E4","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
17:46:42.714 MQT: stat/tasmota_8487E4/STATUS6 = {"StatusMQT":{"MqttHost":"10.0.0.5","MqttPort":1883,"MqttClientMask":"outlet1","MqttClient":"outlet1","MqttUser":"DVES_USER","MqttCount":309,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
17:46:42.719 MQT: stat/tasmota_8487E4/STATUS7 = {"StatusTIM":{"UTC":"2022-01-26T01:46:42","Local":"2022-01-25T17:46:42","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":"-08:00","Sunrise":"23:29","Sunset":"08:35"}}
17:46:42.723 MQT: stat/tasmota_8487E4/STATUS9 = {"StatusPTH":{"PowerDelta":[0,0,0],"PowerLow":0,"PowerHigh":0,"VoltageLow":0,"VoltageHigh":0,"CurrentLow":0,"CurrentHigh":0}}
17:46:42.731 MQT: stat/tasmota_8487E4/STATUS10 = {"StatusSNS":{"Time":"2022-01-25T17:46:42","ENERGY":{"TotalStartTime":"2021-06-30T02:46:10","Total":9.860,"Yesterday":0.077,"Today":0.009,"Power":15,"ApparentPower":20,"ReactivePower":14,"Factor":0.72,"Voltage":117,"Current":0.174}}}
17:46:42.737 MQT: stat/tasmota_8487E4/STATUS11 = {"StatusSTS":{"Time":"2022-01-25T17:46:42","Uptime":"5T02:15:30","UptimeSec":440130,"Heap":22,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":309,"POWER":"ON","Wifi":{"AP":1,REMOVED,"Channel":11,"Mode":"11n","RSSI":66,"Signal":-67,"LinkCount":10,"Downtime":"0T00:21:05"}}}

```
- [X] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_
Deny access to a NTP servers.

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
1. It doesn't spam requests. No point in polling more than once every 5 minutes. Regardless of access.
2. Does not have a ton of brief wifi dropouts.
3. Commands issued to the devices are not delayed.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._
Before (no NTP access) = https://ibb.co/YWymVmk
After (available NTP server assigned) = https://ibb.co/0sfwbzK

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by ascillato** [comment, marked as answer]
There is a bug in the Arduino core that hungs the device for few seconds if there is no NTP, and the same bug produces a reboot exception if the NTP response is not standard.

**Reply by HNJAMeindersma** [reply]
@ascillato's answer selected by @ascillato2.

**Comment by digiblur** [comment]
It is pretty simple and even many routers have local NTP servers available to use.  If you are blocking internet access on your open source ESP chips then you need to be prepared to deal with the fallout that comes with doing so, such as providing local DNS, NTP, and handling your own OTA upgrades and/or OTA server.

**Comment by sfromis** [comment]
If you want to "get by" without real time, you can create a boot rule using the [Time](https://tasmota.github.io/docs/Commands/#time) command to set a fake time, and disable NTP.

**Reply by HNJAMeindersma** [reply]
This should not be the default approach IMO

**Reply by sfromis** [reply]
I disagree. Tasmota is designed for being on a network, and support time-based operation.

**Reply by HNJAMeindersma** [reply]
I agree with your implied design principle, for the network part. But I would say this issue is more about requesting outside resources.

The reason I choose Tasmota above stock or any other firmware is the sense of privacy. (Newer) firmware that tries to connect to anything outside my network with the default configuration would contradict that.

**Reply by HNJAMeindersma** [reply]
NTP is a great functionality to have if you need or want to use something like the (scheduled) timer function. But IMO NTP should not be turned on by default or (if someone has a strong argument for enabling it by default) it should refer to the gateway/router IP by default.

**Reply by sfromis** [reply]
If you do not want to have it access "outside resources", it is real easy to set up a local NTP server. I've done that (in a Docker container), even without worrying about other implications. Many routers and NAS boxes already have this as a feature.

**Reply by HNJAMeindersma** [reply]
Setting up a NTP server is not the (/my) issue here. The issue is that Tasmota requesting external/outside resources (e.g. a NTP server on the WAN) by default. The fact that Tasmota has NTP functionality is not a problem, it's a good thing. Again; the issue is the **default configuration** of the NTP functionality.

**Reply by s-hadinger** [reply]
I don't get you. You would request Tasmota to not request NTP by default? This I would generate so many users complaints.

**Reply by HNJAMeindersma** [reply]
> You would request Tasmota to not request NTP by default?

Yes or 2) to set the NTP servers (`NtpServer1 NtpServer2 NtpServer3`) to the value of the network gateway. Or 3) to let users choose an NTP profile during setup/pairing with the WiFi network.

I think the current default behavior is a spot on the privacy and self-sufficiency reasoning that (I think) brings users to Tasmota.

**Reply by sfromis** [reply]
Your 2) would be pretty bad for a large number of users. If you instead meant DHCP-announced NTP servers, this was tried a long time ago, removed due to often not really working, and commonly ISP routers are not even trying to this, well or badly.

IMO, the number of users seeing NTP server access as a real-world privacy issue is quite limited. "Cloudless" is an argument for Tasmota, but that does not mean NTP, which is pretty easy to deal with for the small minority of users seeing this as a key step. For me, the reason to set up a local NTP server was about not wanting a lot of Tasmota devices "pestering" public NTP servers hourly. Sort of being a "good netizen".

Not having simple onboarding of new devices include a NTP config would be an unhelpful hurdle for the broader Tasmota audience.

**Reply by arendst** [reply]
Totally agree. End of discussion.

**Comment by Jason2866** [comment]
Agree with your first sentence. The requirements to run Tasmota, fully working, are not easy to find (in one place). Since the documentation is Open Source too, you are hereby invited to add this information.
Thank you in advance.

**Comment by lalo-uy** [comment]
You may put an NTP server in the same host of your automation service, and
keep all inside you network.

**Comment by joba-1** [comment]
nice to see I‘m not alone with this wish:
Tasmota should work on the principle of least surprise, which for me means adhere to standards (like getting ntp over dhcp or by manual configuration) and not have unneccessary prerequisites like a set time for relative timers. 
But hey, dealing with the status quo is easier for me and everyone else it seems than to implement this, so I guess it will stay as it is. Tasmota is great anyways.

**Comment by mikerosoft2v** [comment]
Was looking for a solution/workaround today and came across this thread.
I'm using berry script, not the rules engine, so depends on use case.

**Berry timer script working** (working even without getting NTP time after Tasmota restart):

```
# define function:
def tasmotaCmdTelePeriod() tasmota.cmd("TelePeriod") end

# call every second:
# working even without getting NTP time after Tasmota restart.
# taken from timer example: https://tasmota.github.io/docs/Berry-Cookbook/#lvgl-touchscreen-with-3-relays
def tasmotaCmdTelePeriod_every_1s()
  tasmotaCmdTelePeriod()
  tasmota.set_timer(1000, tasmotaCmdTelePeriod_every_1s)
end

# start:
tasmotaCmdTelePeriod_every_1s()
```

Berry cron script I used before (not working without getting NTP time after Tasmota restart):

```
# define function:
def tasmotaCmdTelePeriod() tasmota.cmd("TelePeriod") end

# call every second:
# CAUTION: will not trigger without getting NTP time after Tasmota restart!
tasmota.add_cron("*/1 * * * * *", tasmotaCmdTelePeriod, "tasmotaCmdTelePeriod_every_1s")

# not needed:
#tasmota.remove_cron("tasmotaCmdTelePeriod_every_1s")
```

Version used:
Tasmota v14.3.0

For my use case I need TelePeriod smaller than 3 seconds, and exact timing is not relevant.
The script allows me to continue to use the pre-compiled sml images by ottelo9 for my use case, because I don't want to bother with compiling myself.
(the default minimum TelePeriod is 10 seconds in official pre-compiled images, as well as most third-party).

Using RTC with a battery is not so great. I won't remember (and don't want) to replace the battery after x years.

Regarding previous posts here: Totally understandable that there's always a trade off between maintainability + dev effort vs. changing/adding functionality vs. getting started easily for most setups.

Maybe things changed in Tasmota between versions 10.1.0 and 14.3.0, or only the rules engine timers are limited by this.
Anywho, I'm happy it works for my use case now. 
Tasmota is awesome, thanks!

**Reply by sfromis** [reply]
If you don't care about having correct time, you can use the `Time` command to set a fake current time, and do it after each boot.
